Warn when domain names are not in RFC1035 preferred syntax

mdsahiloss opened the pull request:
<!--
Thank you for contributing to CoreDNS!
Please provide the following information to help us make the most of your pull request:
-->

### 1. Why is this pull request needed and what does it do?

Log a warning when server block zones do not adhere to RFC1035's  "Preferred Syntax"

### 2. Which issues (if any) are related?
This PR fixes #5386

### 3. Which documentation changes (if any) need to be made?

No documentation changes.

### 4. Does this introduce a backward incompatible change or deprecation?
No.

mdsahiloss committed:
Added warning for malformed domain-name

Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -11,6 +12,13 @@ import (
+	// If not correct then shows a warning against domain-name.
+	matchDomain, _ := regexp.Compile("[A-Za-z]+://([A-Za-z0-9-]+\\.[A-Za-z0-9-]+)*:[0-9]+")
+	if !matchDomain.MatchString(addr) {
+		fmt.Println("Warning: Malformed domain-name " + addr)
```
Probably better too append the message to `s` than to print out directly.  Would make it easier to test too.

chrisohaver replied:
Technically it is valid, and not malformed. The message should change to something like ...

`Warning: Domain does not follow RFC1035 preferred syntax`

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -11,6 +12,13 @@ import (
 	s := ""
 
+	// Below block of code checks whether the given domain-name is correct or not.
+	// If not correct then shows a warning against domain-name.
```
replace "correct" with "following RFC1035 preferred syntax"

chrisohaver reviewed (commented):
Please also add a unit test.

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,15 +2,27 @@ package dnsserver
 )
 
+// Below checkDomainName() function checks whether the given domain-name is following RFC1035 preferred syntax or not.
+// If not correct then shows a warning against domain-name.
```
```suggestion
// checkDomainName() returns true if the given domain name follows RFC1035 preferred syntax
```

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,15 +2,27 @@ package dnsserver
 	s := ""
 
+	if !checkDomainName(addr) {
+		s += fmt.Sprintln("Warning: Domain does not follow RFC1035 preferred syntax")
```
Should include the offending name in the warning.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,9 @@
+func TestCheckDomainName(t *testing.T) {
+	if !checkDomainName("dns://example.com:53") {
+		t.Fatalf(`Something went wrong, Now domain name doesn't match with Regular-Expression`)
+	}
```
There should several positive and negative test cases.  i.e. some cases that you expect to get a true result, and some that expect a false result.  

The test failure message should describe the test case and the expected result.  
e.g. `Expected true for "an-example.com"`
or `Expected false for "example-?&^%$.com"`

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,15 +2,27 @@ package dnsserver
 func startUpZones(protocol, addr string, zones map[string]*Config) string {
 	s := ""
 
+	if !checkDomainName(addr) {
```
`addr` is the wrong thing to validate here.  

Each zone key in `zones` needs to be validated.

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,15 +2,27 @@ package dnsserver
+// Below checkDomainName() function checks whether the given domain-name is following RFC1035 preferred syntax or not.
+// If not correct then shows a warning against domain-name.
+func checkDomainName(addr string) bool {
+	matchDomain, _ := regexp.Compile("[A-Za-z]+://([A-Za-z0-9-]+\\.[A-Za-z0-9-]+)*:[0-9]+")
```
This regex should only match a domain name, not the protocol prefix and port suffix.

mariuskimmina commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,12 @@
+	}
+	if !checkDomainName("example-?&^%$.com") {
+		t.Fatalf(`Expected false for "example-?&^$.com"`)
+	}
```
Why do both test cases have an `!`? 
I think something is wrong here.

mariuskimmina replied:
The second test should fail if `checkDomainName` returns true, you have to remove the `!` there. Also your `checkDomainName` function does in fact return true in that second case, which is unintended I think.

mdsahiloss replied:
This is because I'm directly sending a domain-name as an argument to `checkDomainName(domainName)` function and it returns `true` value when domain-name matches Regex and i think we don't need to show  `Expected true for "an-example.com"` output when domain-name matches expected Regex.
whereas, when we pass an wrong `domain-name` as an argument then `checkDomainName(domainName)` returns `false` and I think at that It's needed to print `Expected false for "example-?&^$.com"` that's why I think putting a Not `!` before `checkDomainName(domainName)` gives exprected result.

mariuskimmina replied:
But there is a problem, at the moment your function `checkDomainNames` returns `true` in both cases, which is is unintended and your test is not catching that.

I tried your function in a test program
```
package main

import (
        "fmt"
        "regexp"
)

func checkDomainName(domainName string) bool {
        matchDomain, _ := regexp.Compile("([A-Za-z0-9-]+\\.[A-Za-z0-9-]+)*")
        return matchDomain.MatchString(domainName)
}

func main() {
    output := checkDomainName("an-example.com")
    fmt.Println(output)
    output = checkDomainName("example-?&^%$.com")
    fmt.Println(output)
}
```

And it get the following output 

```
true
true
```

mdsahiloss replied:
Thanks for letting me know,As I can see there is an issue with the Regex so I think I need to improve it.

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,9 +2,16 @@ package dnsserver
 
+// checkDomainName() returns true if the given domain name follows RFC1035 preferred syntax.
+func checkDomainName(domainName string) bool {
+	matchDomain, _ := regexp.Compile("([A-Za-z0-9-]+\\.[A-Za-z0-9-]+)*")
```
Pretty sure this will match a compliant substring in a non-compliant domain name.  
Need to anchor the beginning and end of the regex.  
also pretty sure domain always ends in a dot. 

Please compile coredns with your changes and test both compliant and non-compliant domains. 
Easier / quicker to see if your code works than to iterate with reviews.

mdsahiloss replied:
@chrisohaver  How should I anchor beginning and ending of a domain-name?, I think ending should be anchored by a single `:` but what about beginning.

chrisohaver replied:
https://www.regular-expressions.info/anchors.html

mdsahiloss replied:
Thanks for sharing the docs link, want to ask that, Are you sure? you want a dot `.` at the ending of Regex as ending anchor .

chrisohaver replied:
Two unrelated things ...

1. The regex needs to be anchored at both ends, otherwise it could successfully match a substring of an invalid domain.  e.g. `#!@.example.com.`, would match your current regex because the substring `example.com` matches.
2. Not related to regex anchoring: the variable `zone` will always end in a dot (`.`) because it is a fully qualified domain name (normalized earlier in execution).


(This will be less mysterious if you compile your changes, execute the resulting coredns binary with various  compliant and non-compliant domains, and observe the output)

mdsahiloss replied:
I have improved(anchored beginning and ending) Regex as per your suggestion, means there must be a dot `.` at the ending of Regex for an example if `zone == an-example.com.` then `checkDomainName()` will return true or else will return false although `zone == an-example.com`. Let me know if there is still any issue

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,9 +2,16 @@ package dnsserver
 
+// checkDomainName() returns true if the given domain name follows RFC1035 preferred syntax.
+func checkDomainName(domainName string) bool {
+	matchDomain, _ := regexp.Compile("^[A-Za-z0-9-]{1,}\\.[A-Za-z0-9-]{1,}\\.$")
```
This will fail to match the root domain, `.`. 

Can you share the Corefile you are trying with?

mdsahiloss replied:
okay.. let me update Regex for root domain too.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,12 @@
+import "testing"
+
+func TestCheckDomainName(t *testing.T) {
+	if !checkDomainName("an-example.com") {
```
Won’t this fail with your latest commit? There is no trailing dot in the tested input.

mdsahiloss replied:
I think, this will fail, I need to update test too.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,12 @@
+	if !checkDomainName("an-example.com") {
+		t.Fatalf(`Expected true for "an-example.com"`)
+	}
+	if !checkDomainName("example-?&^%$.com") {
```
This logic is backwards.

mariuskimmina commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,12 @@
+	}
+	if !checkDomainName("example-?&^%$.com") {
+		t.Fatalf(`Expected false for "example-?&^$.com"`)
+	}
```
I think this should really be changed to something like:

```
if checkDomainName("example-?&^%$.com.") {
    t.Fatalf(`Expected false but received true for "example-?&^$.com."`)
}
```

mdsahiloss replied:
Got it.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,14 @@
+	if !checkDomainName("an-example.com.") {
+		t.Fatalf(`Expected true but received false for "an-example.com"`)
+	}
+	if checkDomainName("example-?&^%$.com") {
```
Add a trailing dot to this test.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,14 @@
+	if checkDomainName("example-?&^%$.com") {
+		t.Fatalf(`Expected false but received true for "example-?&^$.com"`)
+	}
+}
```
You need more test cases here.  At the very least you should add the root test case, since your regex didn't account for it initially.

chrisohaver commented on coremain/version.go:
```diff
@@ -2,7 +2,7 @@ package coremain
 
```
This file should not be here.  Please update your master and rebase.

chrisohaver commented on notes/coredns-1.9.3.md:
```diff
@@ -0,0 +1,31 @@
++++
```
This file should not be here.  Please update your master and rebase.

chrisohaver reviewed (changes requested):
This code always crashes.

It is expected for you to compile and test your changes manually and also run your unit tests before submitting PRs for review.  This is a base expectation of all contributors to open source projects.

Please let me know if you need help or guidance on how to build coredns with your changes.

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,9 +2,16 @@ package dnsserver
 
+// checkDomainName() returns true if the given domain name follows RFC1035 preferred syntax.
+func checkDomainName(domainName string) bool {
+	matchDomain, _ := regexp.Compile("[A-Za-z0-9-]{1,}\\.[A-Za-z0-9-]{1,}\\.$)|^\\.$")
```
Suggestion: It would be easier to first check for the root domain case separately than to figure out how to do it in regex.  Also, it would result in a shorter and easier to read regex.

Also, There is no benefit to compiling the regex in a separate step since it's only being evaluated once.  You could instead use single line [regexp.MatchString](https://pkg.go.dev/regexp#MatchString).  To avoid re-compiling the regex for every zone, the compile step needs to be done outside of the zone keys loop in `startUpZones`.

mdsahiloss replied:
I think so, there is no benefit to compile the regex but if I use [regexp.MatchString](https://pkg.go.dev/regexp#MatchString), code will adjust in two steps itself like now, first step to handle return values from `regexp.MatchString()` like `bool` and `err` and then result will be returned in the next step so I think this is good to go If you feel any issue with it so let me know I will update it. 

and your suggestion to check root-domain separately looks good and will be easier for anyone to understand regex, I will update the regex again to make it better.

chrisohaver replied:
> ... I think this is good to go ...

I think the compile step should be done outside of the zone keys loop in startUpZones.

mdsahiloss replied:
As we know `zones` is an Array of Strings, where right now only one string is present (That's root domain-name `.`), But in future when codebase will be deployed , There will be more than one domain-name in `zones` so for that we will need to
call `checkDomainName()` in a for-loop so that we can validate each `zone` of `zones` and now, we are doing the same if I call `checkDomainName()` outside of zone keys loop I'll have to create another for-loop in `startUpZones`.
Let me know if I'm wrong here or you have different vision for it. I'll enhance code as per your suggestion.

chrisohaver replied:
@MdSahil-oss, 

> As we know `zones` is an Array of Strings ...

This is incorrect. `zones` is a map of `string` to `Config`, not an array.  `keys` is a [slice](https://www.godesignpatterns.com/2014/05/arrays-vs-slices.html) of domain names derived from the keys of the `zones` map.

> where right now only one string is present

That's not correct.  It can hold any number of strings.

It is inefficient to re-compile the regex for each domain string in the list.  It should be compiled once, and then the resulting compiled regex can be used to evaluate each domain.  It's not a major problem, since this only happens at startup, and the list of domain names will usually be short.  But IMO it's a good habit to be efficient in code when doing so is easy. Since it's easy to do here, it should be done.

>  I'll have to create another for-loop in startUpZones

No, you don't need to do that.

Please test your code using the Corefile below.  Put temporary print statements in the code (or use a debugger if you already have set up) to look at the value of `keys` in `startupZones`.

This is one Corefile, not 3.
```
# these zones are ok
example.com abc-123.com abc-123 abc.123-xyz a----3 {
 log
}

# these zones should display warnings
abc-^%$.example.com abc-.example.com 123-abc.example.com {
 log
}

# this zone is ok
. {
  log
}
```

mdsahiloss replied:
only `abc-123` and `a----3` are failing from
> # these zones are ok
example.com abc-123.com abc-123 abc.123-xyz a----3 {
 log
}

and others are correct

chrisohaver replied:
They fail because your regex requires at least two labels to be present.

mdsahiloss replied:
@chrisohaver should I improve Regex for single label either ?

chrisohaver replied:
A single label is permitted in the preferred syntax. So yes.

mdsahiloss replied:
Okay... so I'm improving my Regex for single label too but don't forget there will be a dot `.` required as trailing characters for each `zone`. e.g
# these zones are ok
example.com. abc-123.com. abc-123. abc.123-xyz. a----3. {
 log
}

# these zones should display warnings
abc-^%$.example.com. abc-.example.com. 123-abc.example.com. {
 log
}

# this zone is ok
. {
  log
}

chrisohaver replied:
Ok. Feel free to use the regex I [provided](https://github.com/coredns/coredns/pull/5414#issuecomment-1151114989) earlier. You’ll still need to check for the root domain separately, but I think its ok to do that.

chrisohaver reviewed (changes requested):
To summarize the issues that need to be addressed before this can merge:
1. The regex is invalid and does not compile successfully. This causes the subsequent code to panic.
2. The regex doesn't accurately match the RFC 1035 preferred syntax rules. In particular, 1035 has specific rules about permitted leading and trailing characters in each label that the regex does not enforce. 
3. The regex compilation should occur once, per server, and the resulting regex object re-used to match each zone in the server.
4. New test cases need to be added to better cover special cases such as disallowed leading/trailing characters, the root domain, and false positives on preferred-syntax substrings in a non-preferred-syntax domain.
5. Commit history needs to be cleaned up(remove unrelated commits from the PR). Rebasing with updated master should fix it up.

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

Improves Regexp and adds some extra test cases

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,46 @@
+	if regexpObj.MatchString("-example.com.") {
+		t.Fatalf(`Expected true but received false for "-example.com."`)
+	}
+	if regexpObj.MatchString("example.com") {
```
Message and logic mismatch.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,46 @@
+		t.Fatalf(`Expected false but received true for "example-?&^$.com."`)
+	}
+	if regexpObj.MatchString("-example.com.") {
+		t.Fatalf(`Expected true but received false for "-example.com."`)
```
Expected false, right?

mdsahiloss replied:
yeah! right

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,14 +2,22 @@ package dnsserver
 )
 
+// compileRegexp() returns regexp-object for RFC1035 preferred syntax.
+func compileRegexp() *regexp.Regexp {
```
Function name too generic: use something like regex1035PrefSyntax()

chrisohaver reviewed (changes requested):
Should check for reverse zones per earlier comment above. Add test case for that too. 

Please also provide a copy-paste sample of what the output actually looks like when running coredns with multiple zones where some get the warning.

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

Added reverse-domain-name conditions

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -43,4 +49,12 @@ func TestCheckDomainName(t *testing.T) {
+	}
+	if dnsutil.IsReverse("com.example.") == 0 {
+		t.Fatalf(`Expected true for "reverse domain name" but received false for "com.example."`)
+	}
```
In these tests, you should be testing `regexpObj.MatchString` ... not `dnsutil.IsReverse`.

mdsahiloss replied:
ohh! you mean I should remove these tests
~~~ txt
if dnsutil.IsReverse("example.com.") != 0 {
		t.Fatalf(`Expected false for "reverse domain name" but received true for "example.com."`)
	}
if dnsutil.IsReverse("com.example.") == 0 {
t.Fatalf(`Expected true for "reverse domain name" but received false for "com.example."`)
       }
~~~

chrisohaver replied:
Well not exactly.  Per the code comment, your intent here was to add test for reverse zone cases.  That is a good intent, and there should be tests for reverse zones.  However the test cases check `dnsutil.IsReverse`, not your function, also the zones tested are not reverse dns zones (reverse zones doesn't mean reverse order of domain labels).  So, reverse zone cases need to exist, but they need to be fixed.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -24,15 +30,15 @@ func TestCheckDomainName(t *testing.T) {
 	}
 
-	// Below conditions not checks for RFC 1035 "preferred syntax"
+	// Below conditions not checks for RFC 1035 "preferred syntax".
```
This comment is confusing, and I think unnecessary.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -1,16 +1,22 @@
+	"github.com/coredns/coredns/plugin/pkg/dnsutil"
+)
 
 func TestCheckDomainName(t *testing.T) {
```
Function name should contain the function name being tested, e.g. `TestRegex1035PrefSyntax`

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -1,16 +1,22 @@
+	regexpObj := regex1035PrefSyntax() // regexpObj initialised and declared with a regexp-object for RFC1035 preferred syntax.
 
-	// Below conditions checks for RFC 1035 "preferred syntax"
+	// Below condition checks for root domain-name.
```
Comment not necessary. The test case is obvious.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -1,16 +1,22 @@
 		t.Fatalf(`Expected true but received false for "." (root domain)`)
 	}
+
+	// Below conditions checks for RFC 1035 "preferred syntax".
```
This comment is pretty much redundant.

Instead, group the tests into two categories, "positive cases" (where you expect to get a match) and "negative cases" (where you don't expect to get a match).

chrisohaver reviewed (changes requested):
Need to add tests for reverse zone queries.

For clarification on what reverse dns domains are: https://en.wikipedia.org/wiki/Reverse_DNS_lookup

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

Improved Regexp for trailing character

mdsahiloss committed:
Merge branch 'coredns:master' into Warn-of-potentially-malformed-domains-in-Corefiles

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,14 +2,23 @@ package dnsserver
 
+// regex1035PrefSyntax() returns regexp-object for RFC1035 preferred syntax.
+func regex1035PrefSyntax() *regexp.Regexp {
+	matchDomain, _ := regexp.Compile("(^([A-Za-z]([A-Za-z0-9-]*[A-Za-z0-9])(\\.([A-Za-z0-9-]+))*\\.)$)|^(\\.)$")
```
I think it's cleaner to check for the root domain separately. This will simplify the regex (reducing nested parentheses), and make it easier to read.
Also if you use back tics to quote the string, you don't need to escape the `\.`s.  This will also make it easier to read.

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

spliting Regex for root-domain and other-domain separately.

mdsahiloss committed:
Merge remote-tracking branch 'upstream/master' into Warn-of-potentially-malformed-domains-in-Corefiles
merging to keep my Repo updated with upstream master

mdsahiloss committed:
Merge remote-tracking branch 'origin/Warn-of-potentially-malformed-domains-in-Corefiles' into Warn-of-potentially-malformed-domains-in-Corefiles
merging origin/Warn-of-potentially-malformed-domains-in-Corefiles to the Warn-of-potentially-malformed-domains-in-Corefiles branch

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -20,6 +31,9 @@ func startUpZones(protocol, addr string, zones map[string]*Config) string {
 	sort.Strings(keys)
 
 	for _, zone := range keys {
+		if dnsutil.IsReverse(zone) == 0 && !rootDomainRegex.MatchString(zone) && !otherDomainRegex.MatchString(zone) {
```
It would be much simpler (and more efficient) to just do a string compare to check for the root zone instead of   using a regex. i.e.

```
		if zone != "." && dnsutil.IsReverse(zone) == 0 && !otherDomainRegex.MatchString(zone) {
```

chrisohaver reviewed (changes requested):
Since some of your logic is in the `startUpZones` function, you should also be testing that the output of `startUpZones` is as expected.  This should be done in a new test function `TestStartUpZones`. In your test when passing values to the function, the `*Config`s in the `zones` map can be `nil`, since they are not used by `startUpZones.

mdsahiloss committed:
Merge remote-tracking branch 'upstream/master' into Warn-of-potentially-malformed-domains-in-Corefiles
merging to keep the local branch upto date with upstream/master

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

improving as per chrisohaver suggestions

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,15 +2,23 @@ package dnsserver
 
+// checkZoneSyntax() checks whether the given string match 1035 Preferred Syntax or not.
+func checkZoneSyntax(zone string) bool {
+	regex1035PreferredSyntax, _ := regexp.MatchString(`^([A-Za-z][A-Za-z0-9-]*[A-Za-z](\.[A-Za-z][A-Za-z0-9-]*[A-Za-z])*\.)$`, zone)
```
This regex looks more complex than it needs to be - and also looks wrong (doesn't match preferred 1035 syntax).  See the example regex I provided earlier.

Also, the regex match should not be executed if the zone is root or a reverse zone (since you can return true in those cases without needing to evaluate the regex).

mdsahiloss replied:
I think you are right here, I need to improve Regex also for trailing digits as per 1035 syntax.

mdsahiloss replied:
the regex match should not being executed if the zone is root or a reverse zone.

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,45 @@
+		t.Fatalf(`Expected false but received false for "-example.com."`)
+	}
+	if checkZoneSyntax("example.com") {
+		t.Fatalf(`Expected false but received false for "example.com"`)
```
"Expected false but received false" ?

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,45 @@
+		t.Fatalf(`Expected false but received true for "example-?&^$.com."`)
+	}
+	if checkZoneSyntax("-example.com.") {
+		t.Fatalf(`Expected false but received false for "-example.com."`)
```
"Expected false but received false" ?

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,45 @@
+	if !checkZoneSyntax("abc-123.com.") {
+		t.Fatalf(`Expected true but received false for "abc-123.com."`)
+	}
+	if !checkZoneSyntax("abc.123-xyz.") {
```
incorrect expected result.  `abc.123-xyz.` should be _false_, because it has a label that starts with a non-letter.

chrisohaver reviewed (changes requested):
Still needs reverse zone test cases.

Also needs test cases for compliant and non compliant single label zones (e.g. `example.` == true, `example:.` == false), and single character labels (e.g. `a.example.com.` == true, `1.example.com.` == false)

Please carefully review the preferred syntax rules defined in rfc1035 and make sure that:
1. your test cases expected values adhere to that spec (do this first)
2. your regex accurately match that spec

It's important to get the tests right first - otherwise you may end up tweaking your function to match bad tests (I think this has started to happen, re: complexity added to regex to force a bad test pass).

mdsahiloss committed:
Merge remote-tracking branch 'upstream/master' into Warn-of-potentially-malformed-domains-in-Corefiles

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

Improving Regex and adding new testCases

Commit:
log DoH HTTP server error logs in CoreDNS format (#5457)

Signed-off-by: Ondřej Benkovský <ondrej.benkovsky@jamf.com>

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>

fully 1035 syntax following Regex has been set

Commit:
Bump circle-ci image and update README to build using latest go (#5458)

Signed-off-by: Ondřej Benkovský <ondrej.benkovsky@jamf.com>

yongtang committed:
Fix a crash in rewrite plugin when rule type is missing (#5459)

Commit:
plugin/kubernetes : make kubernetes client log in CoreDNS format (#5461)

Signed-off-by: Ondřej Benkovský <ondrej.benkovsky@jamf.com>

yongtang committed:
Fix out-of-index issue in rewrite plugin (#5462)

This PR fixes another out-of-index issue in rewrite to avoid
security vuln.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

dependabot[bot] committed:
build(deps): bump github.com/Azure/azure-sdk-for-go (#5465)

Bumps [github.com/Azure/azure-sdk-for-go](https://github.com/Azure/azure-sdk-for-go) from 65.0.0+incompatible to 66.0.0+incompatible.
- [Release notes](https://github.com/Azure/azure-sdk-for-go/releases)
- [Changelog](https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/Azure/azure-sdk-for-go/compare/v65.0.0...v66.0.0)

---
updated-dependencies:
- dependency-name: github.com/Azure/azure-sdk-for-go
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump github.com/prometheus/common from 0.34.0 to 0.35.0 (#5466)

Bumps [github.com/prometheus/common](https://github.com/prometheus/common) from 0.34.0 to 0.35.0.
- [Release notes](https://github.com/prometheus/common/releases)
- [Commits](https://github.com/prometheus/common/compare/v0.34.0...v0.35.0)

---
updated-dependencies:
- dependency-name: github.com/prometheus/common
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump github.com/aws/aws-sdk-go from 1.44.37 to 1.44.42 (#5467)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.44.37 to 1.44.42.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.44.37...v1.44.42)

---
updated-dependencies:
- dependency-name: github.com/aws/aws-sdk-go
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump k8s.io/klog/v2 from 2.60.1 to 2.70.0 (#5468)

Bumps [k8s.io/klog/v2](https://github.com/kubernetes/klog) from 2.60.1 to 2.70.0.
- [Release notes](https://github.com/kubernetes/klog/releases)
- [Changelog](https://github.com/kubernetes/klog/blob/main/RELEASE.md)
- [Commits](https://github.com/kubernetes/klog/compare/v2.60.1...v2.70.0)

---
updated-dependencies:
- dependency-name: k8s.io/klog/v2
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump google.golang.org/api from 0.84.0 to 0.85.0 (#5469)

Bumps [google.golang.org/api](https://github.com/googleapis/google-api-go-client) from 0.84.0 to 0.85.0.
- [Release notes](https://github.com/googleapis/google-api-go-client/releases)
- [Changelog](https://github.com/googleapis/google-api-go-client/blob/main/CHANGES.md)
- [Commits](https://github.com/googleapis/google-api-go-client/compare/v0.84.0...v0.85.0)

---
updated-dependencies:
- dependency-name: google.golang.org/api
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump github/codeql-action from 2.1.12 to 2.1.14 (#5470)

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 2.1.12 to 2.1.14.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/27ea8f8fe5977c00f5b37e076ab846c5bd783b96...41a4ada31ba866a7f1196b9602703a89edd69e22)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

chrisohaver reviewed (changes requested):
Please address all feedback from prior review(s).

* Regex still longer and more complex than it should be.
* Reverse zone test cases still missing.
* Regex is still evaluated in root and reverse cases.
* etc

chrisohaver committed:
plugin/tsig: new plugin TSIG (#4957)

* expose tsig secrets via dnsserver.Config
* add tsig plugin

Signed-off-by: Chris O'Haver <cohaver@infoblox.com>

Commit:
auto remove trailing whitespaces

Signed-off-by: coredns[bot] <bot@bot.coredns.io>

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
Merge remote-tracking branch 'upstream/master' into Warn-of-potentially-malformed-domains-in-Corefiles
and making some additional changes

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,24 +2,39 @@ package dnsserver
+	if(zone == "." || dnsutil.IsReverse(zone) != 0){
+		return true
+	}
+	regex1035PreferredSyntax, _ := regexp.MatchString(`^([A-Za-z]([A-Za-z0-9-]*[A-Za-z0-9])?(\.[A-Za-z]([A-Za-z0-9-]*[A-Za-z0-9])?)*\.)$`, zone)
```
Your regex will incorrectly match a zone starting with `.` such as `.example.`.
Please use the regex I have suggested. i.e.

```suggestion
	regex1035PreferredSyntax, _ := regexp.MatchString(`^([A-Za-z]([A-Za-z0-9-]*[A-Za-z0-9])?\.)+$`, zone)
```

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,70 @@
+	if checkZoneSyntax("123-abc.example.com.") {
+		t.Fatalf(`Expected false but received true for "123-abc.example.com."`)
+	}
+}
```
```suggestion
	if checkZoneSyntax(".example.com.") {
 		t.Fatal(`Expected false but received true for ".example.com."`)
 	}
}
```

chrisohaver commented on core/dnsserver/onstartup_test.go:
```diff
@@ -0,0 +1,70 @@
+func TestRegex1035PrefSyntax(t *testing.T) {
+
+	// Positive Test cases
+	if !checkZoneSyntax(".") {
```
For all (these) test cases, fail with Error instead of Fatal, so the test will evaluate all cases without aborting on first failure.

Also, you should only use the format version of these functions when you are doing formatting/ value replacement.

Your test zone strings do not always match the zones printed in the error messages ( I suspect because in a couple of places you stripped the `%` from the error message to make the warnings about format strings go away, which was due to your using `Fatalf` instead of `Fatal`).

The copy-paste approach to the test is verbose prone to typo errors.  You should refactor these tests to a loop that checks a series of test cases defined in a slice.  This would reduce copy-paste code replication, making the code shorter, and also less prone to this kind of error. e.g. 

```go
testCases := []struct{
  zone string
  expected bool
}{
  {zone: ".", expected: true},
  {zone: "example.com.", expected: true},
  {zone: "example.", expected: true},
  // etc, add remaining positive and negative test cases here
}

for _, testCase := range testCases {
  if checkZoneSyntax(testCase.zone) != testCase.expected {
    t.Errorf("Expected %v for %q", testCase.expected, testCase.zone)
  }
}
```

(Please note: I just typed the above code example straight into the comment, I didn't check it for correctness, so there may be syntax issues, such as the one I just fixed.)

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
improved Regex and test cases

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
improved Regex and test cases

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,24 +2,39 @@ package dnsserver
 
 	for _, zone := range keys {
+		if !checkZoneSyntax(zone) {
+			s += fmt.Sprintln("Warning: Domain " + zone + " does not follow RFC1035 preferred syntax")
```
```suggestion
			s += fmt.Sprintf("Warning: Domain %q does not follow RFC1035 preferred syntax\n", zone)
```

chrisohaver commented on core/dnsserver/onstartup.go:
```diff
@@ -2,24 +2,39 @@ package dnsserver
+	"github.com/coredns/coredns/plugin/pkg/dnsutil"
 )
 
+// checkZoneSyntax() checks whether the given string match 1035 Preferred Syntax or not.
```
```suggestion
// checkZoneSyntax() checks whether the given string match 1035 Preferred Syntax or not.
// The root zone, and all reverse zones always return true even though they technically don't meet 1035 Preferred Syntax 
```

chrisohaver reviewed (changes requested):
Almost there!

chrisohaver committed:
add tsig to codeowners; update chaos authors (#5479)

Signed-off-by: Chris O'Haver <cohaver@infoblox.com>

Commit:
auto make -f Makefile.doc

Signed-off-by: coredns[bot] <bot@bot.coredns.io>

Commit:
auto go fmt

Signed-off-by: coredns[bot] <bot@bot.coredns.io>

dependabot[bot] committed:
build(deps): bump ossf/scorecard-action from 1.1.1 to 1.1.2 (#5488)

dependabot[bot] committed:
build(deps): bump gopkg.in/DataDog/dd-trace-go.v1 from 1.38.1 to 1.39.0 (#5486)

dependabot[bot] committed:
build(deps): bump github.com/go-logr/logr from 1.2.0 to 1.2.3 (#5485)

dependabot[bot] committed:
build(deps): bump github.com/aws/aws-sdk-go from 1.44.42 to 1.44.47 (#5482)

dependabot[bot] committed:
build(deps): bump github.com/miekg/dns from 1.1.49 to 1.1.50 (#5483)

dependabot[bot] committed:
build(deps): bump github/codeql-action from 2.1.14 to 2.1.15 (#5487)

dependabot[bot] committed:
build(deps): bump google.golang.org/api from 0.85.0 to 0.86.0 (#5484)

mdsahiloss committed:
Update core/dnsserver/onstartup.go

Co-authored-by: Chris O'Haver <cohaver@infoblox.com>

mdsahiloss committed:
Update core/dnsserver/onstartup.go

Co-authored-by: Chris O'Haver <cohaver@infoblox.com>

mdsahiloss committed:
Merge remote-tracking branch 'upstream/master' into Warn-of-potentially-malformed-domains-in-Corefiles

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
Enhanced as per suggestions

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
Enhanced as per suggestions

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
Merge remote-tracking branch 'origin/Warn-of-potentially-malformed-domains-in-Corefiles' into Warn-of-potentially-malformed-domains-in-Corefiles

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
Removed DCO Error

mdsahiloss committed:
Signed-off-by: Md Sahil <Mohdssahil1@gmail.com>
Removed DCO Error

mdsahiloss committed:
Merge remote-tracking branch 'upstream/master' into Warn-of-potentially-malformed-domains-in-Corefiles

mdsahiloss committed:
Merge remote-tracking branch 'origin/Warn-of-potentially-malformed-domains-in-Corefiles' into Warn-of-potentially-malformed-domains-in-Corefiles

Signed-off-by: MdSahil-oss <mohdssahil1@gmail.com>
